doc/synchronization.md

```diff
@@ -0,0 +1,340 @@
+document will provide ground steps for Ruby to become as good in this area as
+in others.
+
+Without a memory model and this layer it's very hard to write concurrent
+abstractions for Ruby. To write a proper concurrent abstraction it often means
+to reimplement it more than once for different Ruby runtimes, which is very
+time-consuming and error-prone.
+
+# Ruby memory model
+
+The Ruby memory model is a framework allowing to reason about programs in
+concurrent and parallel environment. It defines what variable writes can be
+observed by a particular variable read, which is essential to be able to
+determine if a program is correct. It is achieved be defining what subset of
+all possible program execution orders is allowed.
+
+A memory model sources:
+
+-   [Java memory model](http://www.cs.umd.edu/~pugh/java/memoryModel/),
+    and its [FAQ](http://www.cs.umd.edu/~pugh/java/memoryModel/jsr-133-faq.html)
+-   [Java Memory Model Pragmatics](http://shipilev.net/blog/2014/jmm-pragmatics/)
+-   [atomic&lt;&gt; Weapons 1](https://channel9.msdn.com/Shows/Going+Deep/Cpp-and-Beyond-2012-Herb-Sutter-atomic-Weapons-1-of-2)
+and
+[2](https://channel9.msdn.com/Shows/Going+Deep/Cpp-and-Beyond-2012-Herb-Sutter-atomic-Weapons-2-of-2)
+
+Concurrent behavior sources of Ruby implementations:
+
+-   Source codes.
+-   [JRuby's wiki page](https://github.com/jruby/jruby/wiki/Concurrency-in-jruby)
+-   [Rubinius's wiki page](http://rubini.us/doc/en/systems/concurrency/)
+
+> A similar document for MRI was not found. Key fact about MRI is GVL (Global
+> VM lock) which ensures that only one thread can interpret a Ruby code at any
+> given time. When the GVL is handed from one thread to another a mutex is
+> released by first and acquired by the second thread implying that everything
+> done by first thread is visible to second thread. See
+> [thread_pthread.c](https://github.com/ruby/ruby/blob/ruby_2_2/thread_pthread.c#L101-L107)
+> and
+> [thread_win32.c](https://github.com/ruby/ruby/blob/ruby_2_2/thread_win32.c#L95-L100).
+
+This memory model was created by: comparing
+[MRI](https://www.ruby-lang.org/en/), [JRuby](http://jruby.org/),
+[JRuby+Truffle](https://github.com/jruby/jruby/wiki/Truffle),
+[Rubinius](http://rubini.us/); taking account limitations of the implementations
+or their platforms; inspiration drawn from other existing memory models (Java,
+C++11). This is not a formal model.
+
+Key properties are:
+
+-   **volatility (V)** - Any written value is immediately visible to any
+    subsequent volatile read of the same variable. Any write executed before
+    volatile write based on program-order is visible to the read as well, which
+    allows
+    [piggybacking](http://stackoverflow.com/questions/8769570/volatile-piggyback-is-this-enough-for-visiblity).
+    (Same meaning as in Java, it creates synchronizes-with (JMM term) order
+    between write and read, which participates in creating happens-before
+    order.)
```

Comment:
This is very complex now, I think piggybacking is also going too far from the definition.
Maybe we could just say that 'subsequent read' is either same thread, later in the program, or there is some synchronization between the write and the subsequent read.